### PR TITLE
Rewrite to use the new tpmdirect API from go-tpm

### PR DESCRIPTION
**Pull request by Foxboron**

### CODE REVIEW

**Comment by Foxboron** [comment]
@chrisfenner I don't know if you are interested, but here is a rewrite of `age-plugin-tpm` from the legacy tpm2 API to the new `tpmdirect` API.

I'm not sure if this works as some form of feedback or if this helps on the API design :)

**Comment by Foxboron** [comment]
Hrm, actually I tired running this on my actual TPM instead of the swtpm stuff and it fails to do `ECDHZGen`.  Just blocks without any response from the TPM.

File: plugin/plugin.go [hunk 1]
```diff
@@ -27,133 +27,273 @@ const (
...
-	eccKeyParamsDecrypt = tpm2.Public{
-		Type:       tpm2.AlgECC,
-		NameAlg:    tpm2.AlgSHA256,
-		Attributes: tpm2.FlagStorageDefault & ^tpm2.FlagRestricted,
-		ECCParameters: &tpm2.ECCParams{
-			CurveID: tpm2.CurveNISTP256,
-			Point: tpm2.ECPoint{
-				XRaw: make([]byte, 32),
-				YRaw: make([]byte, 32),
+	eccKeyTemplate = tpm2.Create{
```
Comment (chrisfenner) [inline plugin/plugin.go:44]: same as above: I recommend making eccKeyTemplate a TPM2BPublic

**Comment by chrisfenner** [comment]
definitely interested, thanks for tagging me! I have filed a feature request for go-tpm to add a UnmarshalReader/MarshalReader surface, and left a few other comments on the design of your code here. Thanks for using the new go-tpm API!!

File: plugin/key.go [hunk 2]
```diff
@@ -61,10 +61,19 @@ func DecodeIdentity(s string) (*Identity, error) {
...
 	}
 
-	if err := tpmutil.UnpackBuf(r, &key.Private, &key.Public); err != nil {
-		return nil, fmt.Errorf("failed unpacking buffer: %v", err)
+	public, err := tpm2.Unmarshal[tpm2.TPM2BPublic](r.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	private, err := tpm2.Unmarshal[tpm2.TPM2BPrivate](r.Bytes()[len(public.Bytes())+2:])
```
[the remark this answers is not in the document]
Comment (Foxboron) [reply plugin/key.go:69]: Yes, this was the largest annoyance when trying to figure out how to parse the keys :)

**Comment by Foxboron** [comment]
> definitely interested, thanks for tagging me! I have filed a feature request for go-tpm to add a UnmarshalReader/MarshalReader surface, and left a few other comments on the design of your code here. Thanks for using the new go-tpm API!!

Thank you very much for taking the time to review the code :)

File: plugin/key_test.go [hunk 3]
```diff
@@ -4,27 +4,52 @@ import (
...
 	"testing"
 
-	"github.com/google/go-tpm/tpmutil"
+	"github.com/google/go-tpm/tpm2"
 )
 
-func mustPack(d []byte) []byte {
-	packed, err := tpmutil.Pack(d)
+func mustPublic(data []byte) tpm2.TPM2BPublic {
+	tpmdata := tpm2.TPM2BData{
```
[the remark this answers is not in the document]
Comment (Foxboron) [reply plugin/key_test.go:11]:
```
func mustPublic(data []byte) tpm2.TPM2BPublic {
	return tpm2.BytesAs2B[tpm2.TPMTPublic](data)
}
```

Wound up with this. `go-tpm` was shouting at me trying to unmarshal bogus data as `tpm2.TPMTPublic`

File: plugin/key_test.go [hunk 4]
```diff
@@ -4,27 +4,52 @@ import (
...
+	}
+	return *ret
+}
+
+func mustPrivate(data []byte) tpm2.TPM2BPrivate {
+	tpmdata := tpm2.TPM2BData{
+		Buffer: data,
+	}
+	b := tpm2.Marshal(tpmdata)
+	ret, err := tpm2.Unmarshal[tpm2.TPM2BPrivate](b)
```
[the remark this answers is not in the document]
Comment (Foxboron) [reply plugin/key_test.go:27]: Thanks :)

File: plugin/plugin.go [hunk 5]
```diff
@@ -27,133 +27,273 @@ const (
...
+
+	srkHandle, srkPublic, err := CreateSRK(tpm)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed getting srk: %v", err)
+	}
+
+	ecdhRsp, err := ecdh.Execute(tpm,
+		tpm2.HMAC(tpm2.TPMAlgSHA256, 16,
+			tpm2.AESEncryption(128, tpm2.EncryptIn),
```
[the remark this answers is not in the document]
Comment (Foxboron) [reply plugin/plugin.go:278]: Ah, I didn't quite spot that there was a difference. Thanks!

File: plugin/plugin.go [hunk 6]
```diff
@@ -27,133 +27,273 @@ const (
...
+					Buffer: pin,
+				},
+			},
+		}
+		pinstatus = HasPIN
+	}
+
+	var eccRsp *tpm2.CreateResponse
+	eccRsp, err = eccKeyTemplate.Execute(tpm,
+		tpm2.HMAC(tpm2.TPMAlgSHA256, 16,
```
[the remark this answers is not in the document]
Comment (Foxboron) [reply plugin/plugin.go:169]: I was actually expecting it to be a bit more work. The current session tests in `go-tpm` I was basing this off on didn't leave it super obvious if this was the correct way :)

File: plugin/plugin.go [hunk 7]
```diff
@@ -27,133 +27,273 @@ const (
...
-			Symmetric: &tpm2.SymScheme{
-				Alg:     tpm2.AlgAES,
-				KeyBits: 128,
-				Mode:    tpm2.AlgCFB,
-			},
-			CurveID: tpm2.CurveNISTP256,
-			Point: tpm2.ECPoint{
-				XRaw: make([]byte, 32),
-				YRaw: make([]byte, 32),
+	srkTemplate = tpm2.CreatePrimary{
```
[the remark this answers is not in the document]
Comment (Foxboron) [reply plugin/plugin.go:32]: I just moved this from being shared variables to inline them in the function, simpler this way. Thanks!

**Pull request merged**